File: app/src/main/java/com/patloew/countries/data/model/Country.java
package com.patloew.countries.data.model;

import com.patloew.countries.util.RealmListParcelConverter;

import org.parceler.Parcel;
import org.parceler.ParcelPropertyConverter;

import io.realm.CountryRealmProxy;
import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

/* Copyright 2016 Patrick Löwenstein
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License. */
@Parcel(implementations = { CountryRealmProxy.class },
        value = Parcel.Serialization.FIELD,
        analyze = { Country.class })
public class Country extends RealmObject implements Comparable<Country> {

    @PrimaryKey
    public String alpha2Code;
    public String alpha3Code;
    public String name;
    public String nativeName;
    public String region;
    public String capital;
    @ParcelPropertyConverter(RealmListParcelConverter.class)
    public RealmList<RealmString> currencies;
    @ParcelPropertyConverter(RealmListParcelConverter.class)
    public RealmList<RealmString> borders;
    @ParcelPropertyConverter(RealmListParcelConverter.class)
    public RealmList<RealmString> languages;
    @ParcelPropertyConverter(RealmListParcelConverter.class)
    public RealmList<RealmStringMapEntry> translations;
    public Integer population;
    public Float lat;
    public Float lng;

    @Override
    public int compareTo(Country another) {
        if(name != null && another.name != null) {
            return name.compareTo(another.name);
        } else {
            return 0;
        }
    }
}

File: app/src/main/java/com/patloew/countries/data/model/Country.kt
package com.patloew.countries.data.model

import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey
import paperparcel.PaperParcel
import paperparcel.PaperParcelable

/* Copyright 2016 Patrick Löwenstein
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License. */
@PaperParcel
open class Country : RealmObject(), Comparable<Country>, PaperParcelable {

    companion object {
        @JvmField val CREATOR = PaperParcelCountry.CREATOR
    }

    @PrimaryKey
    open var alpha2Code: String? = null
    open var alpha3Code: String? = null
    open var name: String? = null
    open var nativeName: String? = null
    open var region: String? = null
    open var capital: String? = null
    open var currencies: RealmList<RealmString>? = null
    open var borders: RealmList<RealmString>? = null
    open var languages: RealmList<RealmString>? = null
    open var translations: RealmList<RealmStringMapEntry>? = null
    open var population: Int? = null
    open var lat: Float? = null
    open var lng: Float? = null

    override fun compareTo(other: Country): Int {
        if (name != null && other.name != null) {
            return name!!.compareTo(other.name!!)
        } else {
            return 0
        }
    }
}
